### ISSUE TITLE
Question: How to use custom condition matcher?

### ISSUE THREAD

**Issue by dzcpy** [getjerry/nest-casl#135]
According to https://casl.js.org/v5/en/advanced/customize-ability#custom-conditions-matcher-implementation

```ts
import {
  PureAbility,
  AbilityBuilder,
  AbilityTuple,
  MatchConditions,
  AbilityClass
} from '@casl/ability';

type AppAbility = PureAbility<AbilityTuple, MatchConditions>;
const AppAbility = PureAbility as AbilityClass<AppAbility>;
const lambdaMatcher = (matchConditions: MatchConditions) => matchConditions;

export default function defineAbilityFor(user: any) {
  const { can, build } = new AbilityBuilder(AppAbility);

  can('read', 'Article', ({ authorId }) => authorId === user.id);
  can('read', 'Article', ({ status }) => ['draft', 'published'].includes(status));

  return build({ conditionsMatcher: lambdaMatcher });
}
```

Seems like it's much more flexible than `MongoQuery` matcher. However in TypeScript a lamda function cannot be passed to the third argument.
````
No overload matches this call.
  Overload 1 of 2, '(action: string | string[], subject: typeof Team | (typeof Team)[], conditions?: MongoQuery<Team>): RuleBuilder<Ability<AbilityTuple<string, Subject>, MongoQuery<...>>>', gave the following error.
    Type '({ id }: { id: any; }) => boolean' has no properties in common with type 'MongoQuery<Team>'.
  Overload 2 of 2, '(action: string | string[], subject: typeof Team | (typeof Team)[], fields?: string | string[], conditions?: MongoQuery<Team>): RuleBuilder<Ability<AbilityTuple<string, Subject>, MongoQuery<...>>>', gave the following error.
    Argument of type '({ id }: { id: any; }) => boolean' is not assignable to parameter of type 'string | string[]'.
      Type '({ id }: { id: any; }) => boolean' is missing the following properties from type 'string[]': pop, push, concat, join, and 28 more.ts(2769)
```

**Comment by liquidautumn** [comment]
Got it, should be easy to add. I'll take a look at the weekend, will update you on it.

**Comment by liquidautumn** [comment]
Quick update on this one, it wan't easy to add. I had to roll back `Ability` to `PureAbility` changing PR as it break types as is. I tried to find a quick fix, but figured out I'd loose type permissions conditions checks as conditinos type become unknown with `MongoQuery | FunctionalCondition` type. It'd require custom `AbilityBuilder` implementation and updates to `AccessService` and `AbilityFactory` at least. Additionally to handle toSql() conversion of default conditions we'd need to account for lambda conditions type as well.
Technically you probably can still trick TS into accepting lambda conditions by setting PureAbility generic argument to `can` function, but I doubt it'll work down the stack without significant updates.
